config/derive/src/proxy.rs

```diff
@@ -0,0 +1,212 @@
+
+    let proxy_struct_name = &builder_struct.ident;
+
+    let set_field_methods = builder_resolver.set_field_methods;
+    let field_names = builder_resolver.field_names;
+
+    let impl_parent_default = if !builder_resolver.has_required_field {
+        let ident = field_names
+            .iter()
+            .map(|field_value| format_ident!("DEFAULT_{}", field_value.to_string().to_uppercase()));
+
+        quote! {
+            impl Default for #parent_ident {
+                fn default() -> Self {
+                    Configuration::new()
+                }
+            }
+
+            impl #parent_ident {
+                pub const fn new() -> Self {
+                    Self { #( #field_names: Self::#ident ),* }
+                }
+            }
+        }
+    } else {
+        quote!()
+    };
+
+    let builder_default_field_values: Vec<_> = builder_resolver
+        .default_field_values
+        .into_iter()
+        .enumerate()
+        .map(|(i, field_value)| {
+            let field_name = &field_names[i];
+            field_value.map_or_else(
+                || quote! {#field_name: None},
+                |field_value| {
+                    let ident = format_ident!("DEFAULT_{}", field_value.to_string().to_uppercase());
+
+                    quote! {
+                        #field_name: Some(#parent_ident::#ident)
+                    }
+                },
+            )
+        })
+        .collect();
+
+    quote! {
+        #impl_parent_default
+
+        /// [`Configuration`] builder
+        #[derive(
+            Debug, Clone,
+            serde::Deserialize, serde::Serialize,
+            iroha_config_derive::Documented,
+        )]
+        #[serde(default)]
+        #builder_struct
+
+        impl Default for #proxy_struct_name {
+            fn default() -> Self {
+                Self::new()
+            }
+        }
+
+        impl #proxy_struct_name {
+            pub const fn new() -> Self {
+                Self { #( #builder_default_field_values ),* }
+            }
+
+            #( #set_field_methods )*
+
+            /// Override [`Self`] with values from another [`Self`]
+            pub fn override_with(&mut self, other: Self) -> &mut Self { #(
+                if let Some(#field_names) = other.#field_names {
+                    // TODO: Delegate to setter method instead
+                    self.#field_names = Some(#field_names);
+                }; )*
+
+                self
+            }
+
+            /// Build [`Configuration`]
+            ///
+            /// # Errors
+            ///
+            /// - if missing required fields
+            pub fn build(self) -> Result<#parent_ident, ::iroha_config::Error> {
+                Ok(#parent_ident { #(
+                    #field_names: self.#field_names.ok_or(
+                        ::iroha_config::Error::MissingField(stringify!(#field_names))
+                    )? ),*
+                })
+            }
+        }
```

Comment:
but that is only if you use `Self::new` which is equivalent to `Self::default`. You can deserialize from file or from environment and it's possible some fields will be set to `None`

Reply:
I see. But why `Self::new` should be equivalent of `Self::default`? I would prefer to have empty values when I create it with `Self::new()`.

Reply:
> I see. But why Self::new should be equivalent of Self::default? I would prefer to have empty values when I create it with Self::new().

what you say makes sense. `Self::new` should create a struct with empty values. The reason I did this was because `Default::default` cannot be used in const expressions. I will find a different way to do that and make `Self::new` return empty struct

Reply:
should be fixed now